### PR TITLE
🎉 source-facebook-marketing: allow configuration of page_size

### PR DESCRIPTION
**Pull request by mdibaiee**
## What

- Solves #12168

## How
- Allows configuration of page_size 

## Recommended reading order
1. base_streams.py
2. source.py
3. spec.py

## 🚨 User Impact 🚨
No, the default stays 100.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<strong>Updating a connector</strong>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

**Review by sherifnada — CHANGES_REQUESTED** [review]
Is there any way that the connector can automatically adjust page size based on this error? This option will be fairly confusing for users otherwise as were exposing an implementation detail

**Comment by mdibaiee** [comment]
@sherifnada 100 is a magic number here, I don't think there is any explicit reason to use 100 as a fixed number (let me know if there is).

- One solution is to try to be smart about this error, means we need to have a sort of retry logic where the page_size is automatically reduced when this error is hit. This seems to be your suggestion. Imagine a scenario where this happens often (e.g. every request), do we send 2 requests (or more, depending on the adjustment logic) every time? I'm not sure if it's possible to globally change the page_size since each call may be a separate `read` call on the connector.
- Another way is to just reduce this default from 100 to something smaller. I suppose this means more requests need to be sent.
- The other way, which is this PR, is to let the user tweak the number to find something suitable for their data, if need be. Most of the time the default will probably be fine.
- ... any other ideas?

**Comment by mdibaiee** [comment]
To put the issue into perspective, I tried using this branch in our system and we ended up having to reduce the page_size to 25 to get it working without an error, that's 1/4 of the current page_size.

**Comment by mdibaiee** [comment]
@sherifnada I'm wondering if there is any kind of configuration that can be used when interacting with the connectors through Docker, but is not user-facing?

**Comment by sherifnada** [comment]
@mdibaiee this sparked some healthy internal discussion -- I think the right go forward here is to merge your contribution (made some comments on the copy). We'll create a couple of follow-up issues to: 
1. Make this an "advanced" configuration per #3681 
2. Auto-determine this setting inside the connector

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py [hunk 1]
```diff
@@ -147,3 +147,12 @@ class Config:
             "A list which contains insights entries, each entry must have a name and can contains fields, breakdowns or action_breakdowns)"
         ),
     )
+
+    page_size: Optional[PositiveInt] = Field(
+        title="Page Size of Requests",
+        order=7,
+        default=100,
+        description=(
+            "Page size used when sending requests to Facebook API to specify number of records per page when response has pagination."
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/spec.py:156]:
```suggestion
            "Page size used when sending requests to Facebook API to specify number of records per page when response has pagination. Most users do not need to set this field unless they specifically need to tune the connector to address specific issues or use cases."
```

**Review by sherifnada — COMMENTED** [review]
@mdibaiee could you copy-paste the checklist for updating a connector from [here](https://github.com/airbytehq/airbyte/blob/master/.github/pull_request_template.md) into the PR description and fill it out for this contribution? Also if you can sign the CLA we can move forward on this!

**Comment by sherifnada** [comment]
Also @mdibaiee I'm curious -- how do you plan on setting this value in Estuary? Are you expecting it to be an operational fix e.g: if it starts failing for a user, an operator would go and reset this for the user? Or are you exposing it to the end user directly and expecting them to fix on their own?

**Comment by mdibaiee** [comment]
@sherifnada At the moment, a mix of both: we operate some deployments ourselves manually at the moment, but once our UI launches, this will be exposed to users along with all other parameters.

We also understand and share the concern here that this option is not something we want the user to care about, but at this point the work required to hide this away seems more than we can commit to at the moment.

**Comment by mdibaiee** [comment]
I ran the unit tests and they pass. The integration tests however, do not pass with my personal credentials, but it appears to me that it's because there are certain expectations about the data in the account (e.g. certain number of deleted stuff). The logs for integration tests also include my credentials so I didn't paste it.

**Comment by mdibaiee** [comment]
I'll send a fix for that integration test failure, I know what it is.

**Comment by marcosmarxm** [comment]
@sherifnada can you check @mdibaiee changes and see if we can merge it?

**Review by sherifnada — APPROVED** [review]

**Comment by harshithmullapudi** [comment]
/publish connector=connectors/source-facebook-marketing auto-bump-version=false

**Pull request merged**